Section 3's pull request

DvdHalliday opened the pull request:
Console outputs

Exercise 1
![e1](https://user-images.githubusercontent.com/109089386/178785056-c249b56d-d97a-4ebb-ad14-fe9e532b3150.png)
Exercise 2
![e2](https://user-images.githubusercontent.com/109089386/178785042-78efd34e-8c41-4bbc-9356-04af5de27f1d.png)
Exercise 3
![e3](https://user-images.githubusercontent.com/109089386/178785044-78974100-54df-4d5d-b4cb-396a0813cacd.png)
Exercise 4
![e4](https://user-images.githubusercontent.com/109089386/178785046-71f7b1b8-18b3-4031-9d6a-ff60c6408c5f.png)
Exercise 5
![e5](https://user-images.githubusercontent.com/109089386/178785048-74c38deb-5a28-4ecc-bff8-b7971b521a5c.png)
Exercise 6
![e6](https://user-images.githubusercontent.com/109089386/178785050-9255fb26-011f-4ddb-83b8-791b43b0e776.png)
Exercise 7
![e7](https://user-images.githubusercontent.com/109089386/178785051-bc63362c-e75d-4048-a864-1019cb27eb16.png)
Exercise 8
![e8](https://user-images.githubusercontent.com/109089386/178785053-19cf4f76-7478-44aa-ab46-f189ccb98759.png)
Exercise 9
![e9](https://user-images.githubusercontent.com/109089386/178785055-36d3fbdd-87ce-4bbe-9245-f1c1eba3d549.png)

DvdHalliday committed:
Added some stuff to .gitignore to avoid adding .vcxproj  files

DvdHalliday committed:
Update .gitignore

DvdHalliday committed:
First exercise cpp

DvdHalliday committed:
exerrr2

DvdHalliday committed:
exercises 3 n 4

DvdHalliday committed:
exercices 5 and some of 6

DvdHalliday committed:
sixth fully working

DvdHalliday committed:
Fixes and uploads

fixed several exercises and added some others

DvdHalliday committed:
Last update

Prior to reviews, added the instructions of every exercise

DvdHalliday committed:
whoops

DvdHalliday committed:
Whoops

last fix on 8th

DvdHalliday committed:
whoops2

DvdHalliday committed:
whoops3

mcdrakteravision commented:
@DvdHalliday try to commit files that you are fully aware are working correctly and if you need to upload work-in-progress files, then it should be stated on the commit's title.

JDAlzate commented:
Especially mind your commit names. Whoops tells me nothing about what the commit is or what the changes are supposed to be. @DvdHalliday

JDAlzate commented on Section 3/e2/Source.cpp:
```diff
@@ -0,0 +1,50 @@
+			}
+		}
+	} while (!std::cin.good() || (answer != 'T' && answer != 'F'));
+	return getAwardedPointsFromAnswer((answer == 'T') ? 1 : 0,1);
```
Two comments here:
- First, you're using numeric values for booleans instead of directly using true/false, is there any reason for that?
- Second, you now have the same problem you had before. With your logic here, the code will only work if all test answers are true.

jeroo97 replied:
But what if the `actualAnswer` is false? 
And I think it would look cleaner if you pass the value as `true` or `false` instead of `1` or `0`

JDAlzate commented on Section 3/e6/Source.cpp:
```diff
@@ -0,0 +1,58 @@
+
+//Hint: You�ll need to look up stuff here, too.How would you detect invalid inputand an enter key press.
+
+int entered(int amountOfItems, int sum) {
```
Functions are a great tool to prevent code duplication and for documentation within the code. You can pretty much 'name' blocks of code, describing what you're doing just by calling them.

However, functions can make the code harder to read if not handled with care. One of the most important things is that the function's signature should make sense on its own, and it should (as much as possible) not be context-dependent.

Signature: One aspect is the name. 'entered' doesn't really tell me what the function is doing. Another is the return value, why is a function called 'entered' returning an int? Especially given that it's always returning 0.

Context-Sensitive: I can tell that the reason why you're returning a 0 every time is that you intend to call it to exit the main function and return 0 from the main function. However, this responsibility should not be in your function, since now you're forced to use it for this very specific context. You MUST call it in a function that should return 0 and you MUST return the return value of your function. Otherwise, this function returning an int makes no sense, given that it's just supposed to print the result.

So here's my advice:
1. Your functions should very explicitly describe what its purpose is. What about something like 'printResults`?
2. Make functions self-contained. A function should make sense on its own. A problem like this is easily solved my removing the extra responsibility from your function.
Instead of
```
if (std::cin.peek() == '\n') return entered(size, sum);
```
you could do
```
if (std::cin.peek() == '\n')
{
    printResult(size, sum);
    return 0;
}
```

JDAlzate commented on Section 3/e8/Source.cpp:
```diff
@@ -0,0 +1,42 @@
+//Exercise 8 - Easier Lottery
+//Rewrite the previous lottery but now instead of getting $10 for each number matched in the exact same index, give the user $10 for any number that matches anywhere in the sequence.
+
+bool isElementInArray(int element, int array[], int size){
```
Nice! This is a much better function. Very clear what it does and what to expect from it.

JDAlzate commented on Section 3/e8/Source.cpp:
```diff
@@ -0,0 +1,42 @@
+	int lotteryResults[5], lotteryBet[5], winnings = 0, temporaryLotteryResult;
+	std::srand(std::time(0));
+	std::cout << "Welcome to the lottery!\n";
+	//lotteryResults[0] = 2;
```
Avoid leaving commented code in your projects. If you will use it later (most likely you won't), comment specifically why it is commented out and in which situation you would uncomment it. Otherwise, it's useless and you should remove it.

jeroo97 commented on Section 3/e4/Source.cpp:
```diff
@@ -0,0 +1,19 @@
+
+int main() {
+	std::cout << "Multiples of 10:" << std::endl;
+	int sum = 0;
```
I think this variable isn't doing anything. I know it's the same as the previous exercise, but if the code is not serving a purpose, is better not to have it at all. (Same as line 16)

"Lo que no sirve que no estorbe" jeje

JDAlzate commented on Section 3/e9/Source.cpp:
```diff
@@ -0,0 +1,27 @@
+//Do this twice: one time initializing the string with double quotes, and another time initializing it with single characters.
+
+
+void stringReverser(char originalString[],char newString[]) {
```
This one's very nice as well. However, it has the name of an object. Objects are things, so they can be described as nouns. stringReverser represents something that reverses a string.

However, functions are actions - code you wish to execute. In this case, you're *doing* something. So functions should generally be named with a verb as its first word. Something like `reverseString` for instance?

DvdHalliday replied:
reverserOfStrings it is

DvdHalliday replied:
jk

JDAlzate reviewed (changes requested):
Review is up!

DvdHalliday committed:
Review's update

-Added an actualAnswer parameter in ex. 2 so the function getPointsFromQuestion could handle questions that have negative answers.
-Changed several occurrences of true (false) booleans set as 1 (0) in various exercises.
-Removed unnecesary stuff from exercise 4.
-Applied JD's recommendations for exercise 6 regarding the printResult function as well as expanding the if statements in which it was called.
-Removed the useless comments from exercise 8.
-Changed the function name in exercise 9 to reverseString as suggested.

JDAlzate commented on Section 3/e2/Source.cpp:
```diff
@@ -37,14 +37,14 @@ int getPointsFromQuestion(int index, std::string question) {
 		}
 	} while (!std::cin.good() || (answer != 'T' && answer != 'F'));
-	return getAwardedPointsFromAnswer((answer == 'T') ? 1 : 0,1);
+	return getAwardedPointsFromAnswer((answer == 'T') ? true : false, actualAnswer);
```
Better! I missed this in the previous review, but it's a bit verbose to do something like `answer == 'T' ? true : false`.

`if (condition ? true : false)` and `if (condition)` are exactly the same thing, but one of them is longer and redundant. `answer == 'T'` should be more than enough here.

DvdHalliday replied:
true

JDAlzate reviewed (changes requested):
Just one additional comment I missed before hehe. Nice!

DvdHalliday committed:
Redundancy in exercise 2

Removed redundancy in the ternary statement in line 40

JDAlzate reviewed (approved):
Merge approved!
